File: find_duplicate_subtrees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode[]}
 */
var findDuplicateSubtrees = function(root) {
    const res = [];
    const subtrees = {};

    const traverse = function(node) {
        if (!node) {
            return "#";
        }

        const key = node.val + "," + traverse(node.left) + "," + traverse(node.right);

        if (key in subtrees) {
            subtrees[key].push(node);
        } else {
            subtrees[key] = [node];
        }

        if (subtrees[key].length === 2) {
            res.push(subtrees[key][0]);
        }

        return key;
    };

    traverse(root);
    return res;    
};

File: find_duplicate_subtrees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> findDuplicateSubtrees(TreeNode* root) {
        vector<TreeNode*> res;
        unordered_map<string, vector<TreeNode*>> subtrees;

        traverse(root, subtrees);

        for (auto nodes : subtrees) {
            if (nodes.second.size() > 1) {
                res.push_back(nodes.second[0]);
            }
        }

        return res;        
    }

private:
    string traverse(TreeNode* node, unordered_map<string, vector<TreeNode*>>& subtrees) {
        if (node == nullptr) {
            return "#";
        }

        string key = to_string(node->val) + "," + traverse(node->left, subtrees) + "," + traverse(node->right, subtrees);

        subtrees[key].push_back(node);

        return key;
    }
};
